File: civitasJava/src/civitas/Dado.java
package civitas;
import java.util.Random;


/**
 * @file Dado.java
 * @author Yesenia González Dávila
 * @author Esther García Gallego
 * @note Grupo B.3
 * @class Dado
 * @brief Se encarga de todas las decisiones del juego relacionadas con el azar. 
 * @note Sigue el patrón singleton
 */

public class Dado {

    private Random random;                              
    private int ultimoResultado;
    private boolean debug;

    private static Dado instance = new Dado ();
    private static final int SalidaCarcel = 5;

    /** Método de clase para obtener la instancia. */
    static public Dado getInstance () {
        return instance;
    }

    /** Constructor de la clase. */
    Dado () {
        debug = false;
        random = new Random ();                        
        ultimoResultado = -1;                           
    }

    /** Genera un numero aleatorio entre 1 y 6 si el modo debug está desactivado,
     *  o 1 si está activado. */
    int tirar () {
        if (debug)
            ultimoResultado = 1;
        else
            ultimoResultado = random.nextInt(6) + 1;    // Num aleatorio entre 0 y 5, más 1

        return ultimoResultado;
    }

    /** Se tira el dado y se determina si el jugador puede salir de la carcel con el valor obtenido
     * @return @retval true si el numero obtenido le permite salir, @retval false si no es el caso
     */
    boolean salgoDeLaCarcel() {
        boolean salgo = false;
        
        //if (ultimoResultado >= SalidaCarcel)
        if (tirar() >= SalidaCarcel)     
          salgo = true;

        return salgo;
    }

    /** Genera un numero aleatorio entre 0 y n-1 para decidir quien empieza. */
    int quienEmpieza (int n) {
        return random.nextInt(n);                       //Num aleatorio entre 0 y n-1
    }

    /** Activa o desactiva el modo debug. */
    public void setDebug (boolean d) {
        if (d != debug) {                               // Si la configuracion no cambia no se añade el evento a diario
            debug = d;
            String modo = "Modo debug inhabilitado en dado";

            if (debug)
                modo = "Modo debug activo en dado";

            Diario.getInstance().ocurreEvento(modo);
        }
    }

    /** Consultor del atributo ultimoResultado. */
    int getUltimoResultado () {
        return ultimoResultado;
    }
}

File: civitasRuby/pkg/civitasRuby-0.0.1/lib/civitas/Dado.rb
#encoding:utf-8
=begin
Authors: Esther García Gallego
         Yesenia González Dávila
         Grupo B3
=end

require_relative "Diario.rb"

require 'singleton'

module Civitas
  class Dado
    include Singleton
                                  #En ruby no se necesita la instancia
    @@SalidaCarcel = 5               # Los atributos son privados por defecto

    attr_reader :ultimoResultado  #Consultor de ultimoResultado

    def initialize                #Constructor
      @Random
      @ultimoResultado
      @debug = false
    end

    def tirar
      if @debug
        @ultimoResultado = 1
      else
        @ultimoResultado = rand(6) + 1  #Roll a 6 sided die, rand(6) returns a number from 0 to 5 inclusive
      end
    end

    def salgoDeLaCarcel
      salgo = false                     #variable local

      if (tirar >= @@SalidaCarcel)       #E: esto devolverá la variable salgo o no pq es local?
        salgo = true
      else
        salgo = false
      end

    end

    def quienEmpieza (n)
      rand(n)         #E: no dé di hay que incluir algún paquete
    end

    def setDebug (d)
      if d != @debug
        @debug = d
        modo = "Debug desactivado en el dado"
      end

        if @debug
          modo = "Debug activo en el dado"
        end

    Diario.instance.ocurre_evento(modo)
  end
end
end
